### PR TITLE
[Feature] Added MCTSPolicyBase, MCTSPolicy, AlphaGoPolicy, AlphaStarPolicy, and MuZeroPolicy

### PR DESCRIPTION
**Pull request by ParamThakkar123**
## Description

Describe your changes in detail.

Added MCTSPolicyBase, MCTSPolicy, AlphaGoPolicy, AlphaStarPolicy, and MuZeroPolicy to mcts policies.

## Motivation and Context

Why is this change required? What problem does it solve?
If it fixes an open issue, please link to the issue here.
You can use the syntax `close #15213` if this solves the issue #15213

This issue doesn't close but is a subtask of #2357 

- [ ] I have raised an issue to propose this change ([required](https://github.com/pytorch/rl/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X] Documentation (update in the documentation)
- [ ] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [X] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [X] My change requires a change to the documentation.
- [X] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [X] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by ParamThakkar123** [comment]
1 test is currently failing. Working on fixing that

**Comment by ParamThakkar123** [comment]
All tests passed locally

File: torchrl/data/tensor_specs.py [hunk 1]
```diff
@@ -6828,6 +6828,36 @@ def _stack_specs(list_of_spec, dim=0, out=None):
     else:
         raise NotImplementedError
 
+@TensorSpec.implements_for_spec(torch.index_select)
+@Composite.implements_for_spec(torch.index_select)
+def _index_select_spec(input: TensorSpec, dim: int, index: torch.Tensor) -> TensorSpec:
```
Comment (vmoens) [inline torchrl/data/tensor_specs.py]: already implemented below

File: torchrl/modules/mcts/policies.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    """MuZero-style MCTS policy with a specific exploration constant.
+
+    This policy implements the selection from MuZero, using PUCT with c=1.25.
+
+    Args:
+        c (float, optional): Exploration constant. Defaults to 1.25.
+        **kwargs: Additional keyword arguments passed to AlphaGoPolicy.
+    """
+    def __init__(self, *, c: float = 1.25, **kwargs) -> None:
+        super().__init__(c=c, **kwargs)
```
Comment (vmoens) [inline torchrl/modules/mcts/policies.py:160]: This is a 3-class hierarchy to set a single float. 
Let's consider whether class methods or constants would be more appropriate (e.g., MCTSPolicy.from_alphago(), MCTSPolicy.from_muzero())?

Additionally, the naming is somewhat misleading -- AlphaStar and MuZero use substantially different MCTS variants beyond just exploration constant tuning (e.g., MuZero uses a learned model for tree expansion). I guess users might expect these classes to encapsulate those algorithmic differences.

File: torchrl/modules/mcts/policies.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+        Args:
+            c (float, optional): Exploration constant for PUCT. Defaults to 5.0.
+            win_count_key (NestedKey, optional): Key for win counts. Defaults to "win_count".
+            visits_key (NestedKey, optional): Key for visit counts. Defaults to "visits".
+            total_visits_key (NestedKey, optional): Key for total visits. Defaults to "total_visits".
+            prior_prob_key (NestedKey, optional): Key for prior probabilities. Defaults to "prior_prob".
+            action_mask_key (NestedKey, optional): Key for the action mask. Defaults to "action_mask".
+            score_key (NestedKey, optional): Key for storing scores. Defaults to "score".
+            action_key (NestedKey, optional): Key for the selected action. Defaults to "action".
+        """
```
Comment (vmoens) [inline torchrl/modules/mcts/policies.py]: No docstrings in `__init__`

File: torchrl/modules/mcts/policies.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+            while mask.ndim > score.ndim and mask.shape[0] == 1:
+                mask = mask.squeeze(0)
+            if mask.shape != score.shape:
+                mask, score = torch.broadcast_tensors(mask, score)
+            if not mask.any(dim=-1).all():
+                raise ValueError("action_mask must allow at least one valid action per node.")
+            score = torch.where(mask, score, torch.full_like(score, -torch.inf))
+        action = score.argmax(dim=-1)
+        if score.ndim == 1 and action.ndim == 1 and action.numel() == 1:
+            return action.squeeze(0)
```
Comment (vmoens) [inline torchrl/modules/mcts/policies.py]: dead code: `torch.Tensor.argmax(dim=-1)` on a 1D tensor returns a 0-d scalar tensor (ndim=0) (I believe)

File: torchrl/modules/mcts/policies.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+        score_key (NestedKey, optional): Key for storing scores. Defaults to "score".
+        action_key (NestedKey, optional): Key for the selected action. Defaults to "action".
+    """
+    def __init__(self, *, score_module: MCTSScore, action_mask_key: NestedKey = "action_mask", score_key: NestedKey = "score", action_key: NestedKey = "action") -> None:
+        super().__init__(
+            score_module=score_module,
+            action_mask_key=action_mask_key,
+            score_key=score_key,
+            action_key=action_key,
+        )
```
Comment (vmoens) [inline torchrl/modules/mcts/policies.py]: is it any useful?

File: test/test_mcts.py
Comment (vmoens) [inline test/test_mcts.py:1]: Only 3 test methods for the entire new feature:

- No batched input tests -- the policy classes explicitly handle broadcasting and multi-dim masks, but none of this is tested.
- No edge case tests -- e.g., all-False mask (should raise ValueError per the code), different score modules (UCB, EXP3).
- No test for MCTSPolicyBase abstractness -- verify it can't be instantiated directly.
- No test for custom key overrides on the specialized policies.

The existing test changes (`torch.tensor()` removal from assert_close calls) are an unrelated cleanup and should be a separate PR.

File: torchrl/modules/mcts/scores.py [hunk 6]
```diff
@@ -16,6 +16,13 @@
 from tensordict.nn import TensorDictModuleBase
 
 
+class _ScoreFactory:
```
Comment (vmoens) [inline torchrl/modules/mcts/scores.py:19]: Unrelated to this PR?

**Review by vmoens — COMMENTED** [review]
Thanks see my comments

File: torchrl/modules/mcts/scores.py (hunk 6 above)
Comment (ParamThakkar123) [reply torchrl/modules/mcts/scores.py:19]: Yeah this one is related to the scores function PR which gave some errors while executing it with the other policies

File: torchrl/modules/mcts/policies.py (hunk 2 above)
Comment (ParamThakkar123) [reply torchrl/modules/mcts/policies.py:160]: Oh got it @vmoens I will try to refine these implementations that way

**Comment by gtnv** [comment]
hey @ParamThakkar123, is this still being implemented? if not, would you like to close it?

**Comment by ParamThakkar123** [comment]
I can continue to work on it @gtnv. I was waiting on reviews on this. If you can review this I can continue to work on this

**Comment by gtnv** [comment]
if you could update the pr + test e2e that would be great.
